src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp

```diff
@@ -0,0 +1,119 @@
+/// derivatives are not stored, what would be the `MaxDeriv` cannot be updated.
+/// Before any function call, the `QuaternionFunctionOfTime` checks if it has
+/// the same number of stored quaternions as the angular velocity
+/// `PiecewisePolynomial` has stored omegas. This is to ensure we have all the
+/// data necessary to solve the ODE at any time `t`. Other than these few, but
+/// important, differences, the `QuaternionFuntionOfTime` is essentially
+/// identical to the `PiecewisePolynomial`.
+template <size_t MaxDerivReturned>
+class QuaternionFunctionOfTime : public FunctionOfTime {
+ public:
+  QuaternionFunctionOfTime() = default;
+  QuaternionFunctionOfTime(
+      double t, std::array<DataVector, 1> initial_func,
+      gsl::not_null<
+          domain::FunctionsOfTime::PiecewisePolynomial<MaxDerivReturned>*>
+          omega_f_of_t_ptr,
+      double expiration_time) noexcept;
+
+  ~QuaternionFunctionOfTime() override = default;
+  QuaternionFunctionOfTime(QuaternionFunctionOfTime&&) noexcept = default;
+  QuaternionFunctionOfTime& operator=(QuaternionFunctionOfTime&&) noexcept =
+      default;
+  QuaternionFunctionOfTime(const QuaternionFunctionOfTime&) = default;
+  QuaternionFunctionOfTime& operator=(const QuaternionFunctionOfTime&) =
+      default;
+
+  explicit QuaternionFunctionOfTime(CkMigrateMessage* /*unused*/) {}
+
+  auto get_clone() const noexcept -> std::unique_ptr<FunctionOfTime> override;
+
+  // clang-tidy: google-runtime-references
+  // clang-tidy: cppcoreguidelines-owning-memory,-warnings-as-errors
+  WRAPPED_PUPable_decl_template(
+      QuaternionFunctionOfTime<MaxDerivReturned>);  // NOLINT
+
+  void reset_expiration_time(double next_expiration_time) noexcept;
+
+  std::array<double, 2> time_bounds() const noexcept override {
+    return {{stored_quaternions_and_times_.front().time, expiration_time_}};
+  }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) override;
+
+  /// Returns the quaternion at an arbitrary time `t`.
+  std::array<DataVector, 1> func(double t) const noexcept override;
+
+  /// Returns the quaternion and its first derivative at an arbitrary time `t`.
+  std::array<DataVector, 2> func_and_deriv(double t) const noexcept override;
+
+  /// Returns the quaternion and the first two derivatives at an arbitrary
+  /// time `t`.
+
+  std::array<DataVector, 3> func_and_2_derivs(double t) const noexcept override;
+
+  /// Returns the quaternion and the first three derivatives at an arbitrary
+  /// time `t`.
+  std::array<DataVector, 4> func_and_3_derivs(double t) const noexcept;
+
+ private:
+  mutable std::vector<StoredInfo<1>> stored_quaternions_and_times_;
+  domain::FunctionsOfTime::PiecewisePolynomial<MaxDerivReturned>*
+      omega_f_of_t_ptr_;
+  double expiration_time_{std::numeric_limits<double>::lowest()};
+
+  /// Solves the ODE \f$ \dot{q} = \frac{1}{2} q \times \omega \f$ at time `t`
+  /// and stores the result in `quaternion_to_integrate`
```

Comment:
How about "Integrates the ODE \f$ \dot{q} = \frac{1}{2} q \times \omega \f$ from time `t0` to time `t`.  On input, `quaternion_to_integrate` contains the quaternion at time `t0`, and on output, `quaternion_to_integrate` contains the quaternion at time `t`".  (I'm assuming that's where the initial value comes from).